File: backend/internal/model/product.go
package model

import "time"

type Product struct {
	ID            int64     `json:"id"`
	UserID        int64     `json:"user_id"`
	Title         string    `json:"title"`
	Description   string    `json:"description"`
	Price         float64   `json:"price"`
	OriginalPrice *float64  `json:"original_price,omitempty"`
	Category      string    `json:"category"`
	Condition     string    `json:"condition"`
	Images        []string  `json:"images"`
	Contact       string    `json:"contact"`
	Status        int       `json:"status"`
	LikeCount     int       `json:"like_count"`
	CommentCount  int       `json:"comment_count"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`
	IsLiked       bool      `json:"is_liked,omitempty"`
	IsOwner       bool      `json:"is_owner,omitempty"`
}

type CreateProductRequest struct {
	Title         string   `json:"title" binding:"required"`
	Description   string   `json:"description"`
	Price         float64  `json:"price" binding:"required"`
	OriginalPrice *float64 `json:"original_price"`
	Category      string   `json:"category" binding:"required"`
	Condition     string   `json:"condition"`
	Images        []string `json:"images"`
	Contact       string   `json:"contact"`
}

type ProductComment struct {
	ID        int64     `json:"id"`
	ProductID int64     `json:"product_id"`
	UserID    int64     `json:"user_id"`
	Content   string    `json:"content"`
	CreatedAt time.Time `json:"created_at"`
}

type ProductCommentRequest struct {
	Content string `json:"content" binding:"required"`
}

File: frontend/src/api/product.ts
import { request } from "./request"

export interface ProductData {
  id: number
  user_id: number
  title: string
  description: string
  price: number
  original_price?: number
  category: string
  condition: string
  images: string[]
  contact: string
  status: number
  like_count: number
  comment_count: number
  created_at: string
  updated_at: string
  is_liked?: boolean
  is_owner?: boolean
}

export interface ProductComment {
  id: number
  product_id: number
  user_id: number
  content: string
  created_at: string
}

export function createProduct(data: {
  title: string
  description?: string
  price: number
  original_price?: number
  category: string
  condition?: string
  images?: string[]
  contact?: string
}) {
  return request<{ id: number }>({ url: "/products", method: "POST", data })
}

export function listProducts(params?: { category?: string; sort?: string; offset?: number; limit?: number }) {
  return request<ProductData[]>({ url: "/products", method: "GET", data: params as any })
}

export function getProduct(id: number) {
  return request<ProductData>({ url: `/products/${id}`, method: "GET" })
}

export function deleteProduct(id: number) {
  return request<{ message: string }>({ url: `/products/${id}`, method: "DELETE" })
}

export function markSold(id: number) {
  return request<{ message: string }>({ url: `/products/${id}/sold`, method: "PUT" })
}

export function myProducts(params?: { offset?: number; limit?: number }) {
  return request<{ products: ProductData[]; total: number; active: number }>({ url: "/products/my", method: "GET", data: params as any })
}

export function searchProducts(q: string, offset = 0, limit = 20) {
  return request<ProductData[]>({ url: "/products/search", method: "GET", data: { q, offset, limit } })
}

export function toggleProductLike(id: number) {
  return request<{ liked: boolean }>({ url: `/products/${id}/like`, method: "POST" })
}

export function createProductComment(productId: number, content: string) {
  return request<{ id: number }>({ url: `/products/${productId}/comments`, method: "POST", data: { content } })
}

export function listProductComments(productId: number) {
  return request<ProductComment[]>({ url: `/products/${productId}/comments`, method: "GET" })
}

export function productCategories() {
  return request<string[]>({ url: "/products/categories", method: "GET" })
}

export function trendingProducts() {
  return request<ProductData[]>({ url: "/products/trending", method: "GET" })
}
